### PR TITLE
Datastore conn init

### PR DESCRIPTION
**Pull request by klinga**
1. This update provides basic connectivity to postgres via `datastore.DataAccessLayer`. I added a context manager (`session_scope`) to add a convenient method to call and exit a db session. I made assumption that db settings will be pulled from env variables as it is for local and Travis tests.
2. Also created a new module datastore_transactions.py to house any methods to communicate with the db. At the moment there are only two basic operations - I will add more later when I have a better idea what is needed.
3. Made some changes to tables: opted for internal/independent identifier for the Resource table (previously it had two primary keys: sierraId and libraryId), but that somewhat complicated relations between the tables (double foreign keys?). Instead I added a unique constraint for combination of sierraId and libraryId that should accomplish the same thing.
4. Do I need to worry about character encoding in Postgres? Collation? I had some issues in MySQL, but it looks Postgres is handling it better. 

It took me a while to figure out correct Travis settings to install Posgresql 13 (their default is 9.6 which I thought may not be a good idea). Made some silly but frustrating mistakes along the way...

### CODE REVIEW

File: nightshift/datastore.py [hunk 1]
```diff
@@ -24,12 +22,59 @@
...
+        NS_DBHOST=os.getenv("NS_DBHOST"),
+        NS_DBPORT=os.getenv("NS_DBPORT"),
+        NS_DBNAME=os.getenv("NS_DBNAME"),
+    )
+
+
+class DataAccessLayer:
+    def __init__(self):
+        db = conf_db()
+        self.conn = f"postgresql+psycopg2://{db['NS_DBUSER']}:{db['NS_DBPASSW']}@{db['NS_DBHOST']}:{db['NS_DBPORT']}/{db['NS_DBNAME']}"
```
Comment (mwbenowitz) [inline nightshift/datastore.py]: I don't think you need `+psycopg2` here (minor thing to just keep this a simple as possible)

File: nightshift/datastore.py [hunk 2]
```diff
@@ -24,12 +22,59 @@
...
+    Provides a transactional scope around series of operations.
+    """
+    dal.connect()
+    session = dal.Session()
+    try:
+        yield session
+    except:
+        session.rollback()
+        raise
+    finally:
```
Comment (mwbenowitz) [inline nightshift/datastore.py:75]: Would it be a good or bad idea to automatically run `session.commit()` here? Unsure of how granular you want the commits to be. Could also provide a backstop to ensure that no changes go uncommitted to the db.

File: nightshift/datastore_transactions.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    Args:
+        session:                sqlalchemy.Session instance
+        model:                  one of datastore table classes
+        kwargs:                 new record values as dictionary
+
+    Returns:
+        instance of inserted or duplicate record
+    """
+    instance = session.query(model).filter_by(**kwargs).first()
```
Comment (mwbenowitz) [inline nightshift/datastore_transactions.py]: Should these records always be unique? If so `one_or_none()` might be better here

File: nightshift/datastore_transactions.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    if not instance:
+        instance = model(**kwargs)
+        session.add(instance)
+    return instance
```
Comment (mwbenowitz) [inline nightshift/datastore_transactions.py]: Should this return some information as to whether the record was created or not? Not sure the context in which this will be used but I can imagine knowing if a record already exists in the db might prove useful. (This can be learned by checking if the returned record has a `nid` so this isn't strictly necessary but not sure how you feel about relying on that.

File: tests/test_datastore_transactions.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        sourceId=src_rec.nid,
+        status="open",
+    )
+    test_session.commit()
+    assert rec1.status == "open"
+    rec2 = update_resource(
+        test_session, sierraId=sierraId, libraryId=lib_rec.nid, status="expired"
+    )
+    assert rec2.nid == rec1.nid
+    assert rec1.status == "expired"
```
Comment (mwbenowitz) [inline tests/test_datastore_transactions.py:53]: Add a test here to verify that the `update_resource` method returns `None` when it tries to update a non-existent record (and that nothing was changed in the db)?

File: tests/test_datastore.py [hunk 6]
```diff
@@ -84,3 +106,9 @@ def test_WorldcatQuery_tbl_repr():
         )
         == "<WorldcatQuery(nid='1', resourceId='2', libraryId='1', match='False', responseCode='404')>"
     )
+
+
+def test_datastore_connection(test_connection):
+    with does_not_raise():
+        engine = create_engine(test_connection)
+        engine.connect()
```
Comment (mwbenowitz) [inline tests/test_datastore.py:140]: It would be great to add a test for `session_scope` her as well, I think it's the only thing you've added that's not tested.

**Review by mwbenowitz — APPROVED** [review]
This all looks great to me. The overall structure feels very clean and understandable (so my comments are pretty specific haha). I think the decision to swap out the multi-column primary key for an internal key & unique constraint is a good idea.

I have not run into any issues with relying on the default implementation of unicode in postgres. As for collation, that could be an issue, but I'm not sure what your specific use case is. I don't think anything here would pose an issue so far though.

The only thing I am thinking about generally is the two database communication methods you have. I think that with that model we should agree upon a general contract that those methods would use for their returns. Right now each can return a value that essentially means "nothing happened", but in one a record is returned and in the other `None` is returned. I'm not sure but I think that they should return `None` if nothing changed, a record if a new one was created, and something{?) for an update. Generally unsure, but something like that would make communications with the db very clear I think.

File: nightshift/datastore.py (hunk 1 above)
Comment (klinga) [reply nightshift/datastore.py]: Oops, you are right - sqlalchemy uses psycopg2 as a default driver already.

File: nightshift/datastore.py (hunk 2 above)
Comment (klinga) [reply nightshift/datastore.py:75]: I've been thinking about it but could not decide one way or the other. I think there might be situations when it would be better to commit records before session ends, but that can be done explicitly mid session too. I will add commit before session closes.

File: nightshift/datastore_transactions.py (hunk 3 above)
Comment (klinga) [reply nightshift/datastore_transactions.py]: Looked up `one_or_none()` and yes, it will work better here.

File: tests/test_datastore_transactions.py (hunk 5 above)
Comment (klinga) [reply tests/test_datastore_transactions.py:53]: Done.

File: nightshift/datastore_transactions.py (hunk 4 above)
Comment (klinga) [reply nightshift/datastore_transactions.py]: Thanks for the suggestion. I agree, the current approach is not consistent. I like your general idea to return an instance if something has been changed and nothing otherwise. Will try to apply the same rule in other methods in this module.

File: tests/test_datastore.py (hunk 6 above)
Comment (klinga) [reply tests/test_datastore.py:140]: Done.

**Comment by klinga** [comment]
Thanks @mwbenowitz! I considered all of your suggestions and updated accordingly this PR.
Will merge it now.
